src/viz/viewer_renderer.cpp

```diff
@@ -416,6 +416,414 @@ static VkQueue makeGFXQueue(const Device &dev, uint32_t idx)
+    subgroup_size.sType =
+        VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_REQUIRED_SUBGROUP_SIZE_CREATE_INFO_EXT;
+    subgroup_size.pNext = nullptr;
+    subgroup_size.requiredSubgroupSize = 32;
+#endif
+
+    compute_infos[0].sType = VK_STRUCTURE_TYPE_COMPUTE_PIPELINE_CREATE_INFO;
+    compute_infos[0].pNext = nullptr;
+    compute_infos[0].flags = 0;
+    compute_infos[0].stage = {
+        VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO,
+        nullptr, //&subgroup_size,
+        VK_PIPELINE_SHADER_STAGE_CREATE_REQUIRE_FULL_SUBGROUPS_BIT_EXT,
+        VK_SHADER_STAGE_COMPUTE_BIT,
+        shader.getShader(0),
+        "voxelGen",
+        nullptr,
+    };
+    compute_infos[0].layout = voxel_gen_layout;
+    compute_infos[0].basePipelineHandle = VK_NULL_HANDLE;
+    compute_infos[0].basePipelineIndex = -1;
+
+    std::array<VkPipeline, compute_infos.size()> pipelines;
+    REQ_VK(dev.dt.createComputePipelines(dev.hdl, pipeline_cache,
+        compute_infos.size(),
+        compute_infos.data(), nullptr,
+        pipelines.data()));
+
+    FixedDescriptorPool desc_pool(dev, shader, 0, num_frames);
+
+    return Pipeline<1> {
+        std::move(shader),
+            voxel_gen_layout,
+            pipelines,
+            std::move(desc_pool),
+    };
+}
+
+static PipelineShaders makeVoxelDrawShaders(
+    const Device& dev, VkSampler repeat_sampler, VkSampler clamp_sampler)
+{
+    (void)repeat_sampler;
+    (void)clamp_sampler;
+
+    std::filesystem::path shader_dir =
+        std::filesystem::path(STRINGIFY(VIEWER_DATA_DIR)) /
+        "shaders";
+
+    auto shader_path = (shader_dir / "voxel_draw.hlsl").string();
+
+    ShaderCompiler compiler;
+    SPIRVShader vert_spirv = compiler.compileHLSLFileToSPV(
+        shader_path.c_str(), {}, {},
+        { "vert", ShaderStage::Vertex });
+
+    SPIRVShader frag_spirv = compiler.compileHLSLFileToSPV(
+        shader_path.c_str(), {}, {},
+        { "frag", ShaderStage::Fragment });
+
+#if 0
+    {0, 2, repeat_sampler, 1, 0},
+    { 0, 3, clamp_sampler, 1, 0 },
+    { 1, 1, VK_NULL_HANDLE,
+        VulkanConfig::max_materials *
+            VulkanConfig::textures_per_material,
+     VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT },
+#endif
+
+        std::array<SPIRVShader, 2> shaders {
+        std::move(vert_spirv),
+            std::move(frag_spirv),
+    };
+
+    StackAlloc tmp_alloc;
+    return PipelineShaders(dev, tmp_alloc, shaders,
+        Span<const BindingOverride>({
+            BindingOverride {
+                1,
+                0,
+                VK_NULL_HANDLE,
+                InternalConfig::maxTextures,
+                VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT,
+            },
+            BindingOverride {
+                1,
+                1,
+                repeat_sampler,
+                1,
+                0,
+            },
+            }));
+}
+
+static Pipeline<1> makeVoxelDrawPipeline(const Device& dev,
```

Comment:
Does this function have functional differences from the other draw pipeline functions? How hard would it be to combine common functionality?